8361099: Shenandoah: Improve heap lock contention by using CAS for memory allocation

pengxiaolong opened the pull request:
- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).

Shenandoah always allocates memory with heap lock, we have observed heavy heap lock contention on memory allocation path in performance analysis of some service in which we tried to adopt Shenandoah. This change is to propose an optimization for the code path of memory allocation to improve heap lock contention, along with the optimization, a better OOD is also done to Shenandoah memory allocation to reuse the majority of the code:

* ShenandoahAllocator: base class of the allocators, most of the allocation code is in this class.
* ShenandoahMutatorAllocator: allocator for mutator, inherit from ShenandoahAllocator, only override methods `alloc_start_index`, `verify`, `_alloc_region_count` and  `_yield_to_safepoint` to customize the allocator for mutator.
* ShenandoahCollectorAllocator: allocator for collector allocation in Collector partition, similar to ShenandoahMutatorAllocator, only few lines of code to customize the allocator for Collector. 
* ShenandoahOldCollectorAllocator:  allocator for mutator collector allocation in OldCollector partition, it doesn't inherit the logic from ShenandoahAllocator for now, the `allocate` method has been overridden to delegate to `FreeSet::allocate_for_collector` due to the special allocation considerations for `plab` in old gen. We will rewrite this part later and move the code out of `FreeSet::allocate_for_collector`

I'm not expecting significant performance impact for most of the cases since in most case the contention on heap lock it not high enough to cause performance issue, but in some cases it may improve the latency/performance:

1. Dacapo lusearch test on EC2 host with 96 CPU cores, p90 is improved from 500+us to less than 150us, p99 from 1000+us to ~200us. 
```
java -XX:-TieredCompilation -XX:+AlwaysPreTouch -Xms31G -Xmx31G -XX:+UseShenandoahGC -XX:+UnlockExperimentalVMOptions -XX:+UnlockDiagnosticVMOptions  -XX:-ShenandoahUncommit -XX:ShenandoahGCMode=generational  -XX:+UseTLAB -jar ~/tools/dacapo/dacapo-23.11-MR2-chopin.jar  -n 10 lusearch  | grep "metered full smoothing"
```

Openjdk TIP:
```
===== DaCapo tail latency, metered full smoothing: 50% 241098 usec, 90% 402356 usec, 99% 411065 usec, 99.9% 411763 usec, 99.99% 415531 usec, max 428584 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 902 usec, 90% 3713 usec, 99% 5898 usec, 99.9% 6488 usec, 99.99% 7081 usec, max 8048 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 2608 usec, 90% 6860 usec, 99% 8741 usec, 99.9% 9381 usec, 99.99% 9864 usec, max 10868 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 37 usec, 90% 526 usec, 99% 1118 usec, 99.9% 1668 usec, 99.99% 2171 usec, max 3473 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 30 usec, 90% 516 usec, 99% 1046 usec, 99.9% 1624 usec, 99.99% 2112 usec, max 2966 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 31 usec, 90% 514 usec, 99% 1038 usec, 99.9% 1635 usec, 99.99% 2143 usec, max 3341 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 31 usec, 90% 516 usec, 99% 1043 usec, 99.9% 1633 usec, 99.99% 2184 usec, max 3316 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 29 usec, 90% 514 usec, 99% 1043 usec, 99.9% 1644 usec, 99.99% 2130 usec, max 2899 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 31 usec, 90% 516 usec, 99% 1055 usec, 99.9% 1632 usec, 99.99% 2150 usec, max 2801 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 32 usec, 90% 515 usec, 99% 1048 usec, 99.9% 1651 usec, 99.99% 2167 usec, max 3357 usec, measured over 524288 events =====

```
With CAS allocator:
```
===== DaCapo tail latency, metered full smoothing: 50% 286136 usec, 90% 393094 usec, 99% 398962 usec, 99.9% 399416 usec, 99.99% 407037 usec, max 421050 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 90 usec, 90% 10127 usec, 99% 12901 usec, 99.9% 13061 usec, 99.99% 13148 usec, max 22555 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 24 usec, 90% 163 usec, 99% 1586 usec, 99.9% 1810 usec, 99.99% 1961 usec, max 11508 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 24 usec, 90% 160 usec, 99% 1305 usec, 99.9% 1554 usec, 99.99% 2178 usec, max 10064 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 22 usec, 90% 148 usec, 99% 216 usec, 99.9% 853 usec, 99.99% 1003 usec, max 10221 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 22 usec, 90% 147 usec, 99% 209 usec, 99.9% 739 usec, 99.99% 879 usec, max 1419 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 22 usec, 90% 147 usec, 99% 207 usec, 99.9% 717 usec, 99.99% 855 usec, max 970 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 22 usec, 90% 147 usec, 99% 206 usec, 99.9% 715 usec, 99.99% 859 usec, max 977 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 22 usec, 90% 147 usec, 99% 206 usec, 99.9% 695 usec, 99.99% 832 usec, max 949 usec, measured over 524288 events =====
===== DaCapo tail latency, metered full smoothing: 50% 22 usec, 90% 147 usec, 99% 207 usec, 99.9% 695 usec, 99.99% 833 usec, max 949 usec, measured over 524288 events =====

```

Same tests, if I run it on smaller instance with less CPU cores, the improvement is much less. 

2. specJBB2015 on EC2 with 2 cores, 4G heap:

|                    |                     | GenShen - tip | Genshen - CAS allocator | CAS vs TIP |
|--------------------|---------------------|---------------|-------------------------|------------|
| 1st                | Max jOPS            | 2998          | 3108                    | 3.67%      |
|                    | Critical jOPS       | 1489          | 1601                    | 7.52%      |
| 2nd                | Max jOPS            | 2735          | 2669                    | -2.41%     |
|                    | Critical jOPS       | 1533          | 1454                    | -5.15%     |
| 3rd                | Max jOPS            | 3031          | 3187                    | 5.15%      |
|                    | Critical jOPS       | 1582          | 1527                    | -3.48%     |
| Average            | Max jOPS            | 2921.33333    | 2988                    | 2.28%      |
|                    | Critical jOPS       | 1534.66667    | 1527.33333              | -0.48%     |
| Standard deviation | Max jOPS Stdev      | 132.44454     | 227.86107               |            |
|                    | Critical jOPS Stdev | 37.98538      | 60.01296                |            |

3. specJbb2015 result from our performance pipeline:

<img width="1204" height="417" alt="Specjbb2015-128g" src="https://github.com/user-attachments/assets/4e684612-8ec9-4f6b-9d3e-2f53648158ad" />

<img width="1182" height="687" alt="Specjbb2015" src="https://github.com/user-attachments/assets/3d138e6f-fe41-4f68-863b-780f687fb6cc" />

### Other tests:
- [x] All Shenandoah jtreg tests

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8361099](https://bugs.openjdk.org/browse/JDK-8361099): Shenandoah: Improve heap lock contention by using CAS for memory allocation (**Enhancement** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/26171/head:pull/26171` \
`$ git checkout pull/26171`

Update a local copy of the PR: \
`$ git checkout pull/26171` \
`$ git pull https://git.openjdk.org/jdk.git pull/26171/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 26171`

View PR using the GUI difftool: \
`$ git pr show -t 26171`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/26171.diff">https://git.openjdk.org/jdk/pull/26171.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/26171#issuecomment-3317617736)
</details>

kdnilsen reviewed (commented):
Thanks for doing this work.  It is huge progress.  I've left a number of comments.  I didn't have time to study/comment on all of the code, as I am out-of-office for rest of this week.  I'll look more when I return.

pengxiaolong commented on src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:
```diff
@@ -244,21 +244,18 @@ void ShenandoahRegionPartitions::establish_mutator_intervals(idx_t mutator_leftm
   _leftmosts_empty[int(ShenandoahFreeSetPartitionId::Mutator)] = mutator_leftmost_empty;
   _rightmosts_empty[int(ShenandoahFreeSetPartitionId::Mutator)] = mutator_rightmost_empty;
 
```
Yes, we do have heap lock. But mutator may also update the used/available after allocation w/o heap lock, so the update for mutator partition need to be done with atomic operation, but not all of them. 

I'll revisit this part and remove the atomic operations which are not needed.

pengxiaolong commented on src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:
```diff
@@ -244,21 +244,18 @@ void ShenandoahRegionPartitions::establish_mutator_intervals(idx_t mutator_leftm
-  _capacity[int(ShenandoahFreeSetPartitionId::Mutator)] = mutator_region_count * _region_size_bytes;
-  _available[int(ShenandoahFreeSetPartitionId::Mutator)] =
-    _capacity[int(ShenandoahFreeSetPartitionId::Mutator)] - _used[int(ShenandoahFreeSetPartitionId::Mutator)];
+  Atomic::store(_region_counts + int(ShenandoahFreeSetPartitionId::Mutator), mutator_region_count);
```
It make sense, I'll update it,  `[]` operation for array always has boundary check.

earthling-amzn reviewed (commented):
It would be helpful to have a more detailed overview in the change log description to help reviewers get oriented with the changes.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:
```diff
@@ -744,12 +759,60 @@ void ShenandoahRegionPartitions::assert_bounds() {
 }
 #endif
 
+PaddedEnd<ShenandoahDirectlyAllocatableRegionAffinity::Affinity>* ShenandoahDirectlyAllocatableRegionAffinity::_affinity = nullptr;
```
IIUC, each of the DirectlyAllocatableRegions has an affinity to a particular thread.  If some other thread randomly selects to allocate from this region, we change the region's affinity.  Now, if the original thread tries another allocation, it will be redirected to a different randomly selected region.

It feels to me like this introduces more churn than necessary.  It should be ok for multiple threads to have affinity to the same directly allocatable region so they can preserve "locality".  The locality to be preserved is the value of region->top() and region->end(), in local caches.  Sometimes, the multiple threads that are affiliated with a particular region will be running on the same core, which helps improve cache performance.  If they are on different cores, that's now worse than the current implementation.

So my thinking is that each thread should maintain an affinity to a particular index within the directly allocatable region array.  It should reaffiliate with a different region only if that region entry becomes nullptr, or in case its attempt to CAS allocate fails more than twice in a row (because of heavy contention on the value of top for that region).

Please clarify if I've misunderstood.  Please explain rationale if there is good reason for preferring the design as it is currently implemented.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:
```diff
@@ -2095,6 +2167,326 @@ HeapWord* ShenandoahFreeSet::allocate(ShenandoahAllocRequest& req, bool& in_new_
+    }
+    i++;
+  }
+  return obj;
```
I think obj always equals nullptr at this point.  Seems the code would be easier to understand (and would depend less on effective compiler optimization) if we just made that explicit.  Can we just say:

return nullptr?

pengxiaolong replied:
Yes, it is always `nullptr`, `return nullptr` will make the code more readable.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:
```diff
@@ -2095,6 +2167,326 @@ HeapWord* ShenandoahFreeSet::allocate(ShenandoahAllocRequest& req, bool& in_new_
+  uint start_idx = ShenandoahDirectlyAllocatableRegionAffinity::index();
+  const uint max_probes = ShenandoahDirectAllocationMaxProbes;
+  for (;;) {
+    HeapWord* obj = nullptr;
```
This is written as an infinite loop, but I'm not sure it intends to iterate?  It looks like there's no way for the loop body to get to a second iteration.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:
```diff
@@ -2095,6 +2167,326 @@ HeapWord* ShenandoahFreeSet::allocate(ShenandoahAllocRequest& req, bool& in_new_
+  return obj;
+}
+
+class DirectAllocatableRegionRefillClosure final : public ShenandoahHeapRegionIterationClosure {
```
I don't think we want to subclass ShenandoahHeapRegionIterationClosure here.  That iterates over all 2000 regions.  We only want to iterate over the 13 Directly allocatable regions.  Maybe we don't even need/want a closure iterator here.  We could just write a loop.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp:
```diff
@@ -360,25 +361,29 @@ void ShenandoahHeapRegion::make_committed_bypass() {
 }
 
 void ShenandoahHeapRegion::reset_alloc_metadata() {
```
Do we need to make these atomic because we now increment asynchronously from within mutator CAS allocations?  Before, they were only adjusted while holding heap lock?  I'm wondering if add-with-fetch() or CAS() would be more/less efficient than AtomicAccess::stores.  Can we test the tradeoffs?

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.inline.hpp:
```diff
@@ -108,6 +108,89 @@ HeapWord* ShenandoahHeapRegion::allocate(size_t size, const ShenandoahAllocReque
+  return obj;
+}
+
+HeapWord* ShenandoahHeapRegion::allocate_atomic(size_t size, const ShenandoahAllocRequest& req) {
```
As mentioned above, this is maybe where we would want to set the thread-local _index variable, and this is also where we might want to count how many times we retry the try_allocate() request before we succeed.

The point is that if we have multiple try_allocate() CAS failures, that means this region is heavily contended, and we don't want to make this our new "starting index".

pengxiaolong committed:
Address test failures after merging the change from master which unify the accounting in FreeSet and ShenandoahGeneration

pengxiaolong committed:
Add assert back

pengxiaolong committed:
Set region back to empty and unaffiliated when release a directly allocatable region(may happen in full GC)

pengxiaolong committed:
Remove ShenandoahDirectAllocationMaxProbes, simplifications

pengxiaolong committed:
tidy up

pengxiaolong committed:
Remove unnecessary fences

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong committed:
tidy up

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong committed:
Merge branch 'cas-alloc-1' of https://github.com/pengxiaolong/jdk into cas-alloc-1

pengxiaolong committed:
Remove unnecessary test for directly allocable region

pengxiaolong committed:
Fix wrong asserts

pengxiaolong committed:
Remove ShenandoahDirectlyAllocatableRegionAffinity and simply use thread local instead

pengxiaolong committed:
Renaming directly allocatable region to alloc region

pengxiaolong committed:
Refactor atomic alloction for mutator and make the impl easy to extend to collector

pengxiaolong committed:
Use SIZE_MAX instead of SIZE_T_MAX

pengxiaolong committed:
Minor fix

pengxiaolong committed:
Use ShenandoahCollectorAllocator for collector alloc request in young gen

pengxiaolong committed:
Release collector alloc regions at final-mark

pengxiaolong committed:
Fix issues with collector allocation

pengxiaolong committed:
Move set_update_watermark to atomic_allocate_in

pengxiaolong committed:
Use ShenandoahOldCollectorAllocator for old collector allocation

pengxiaolong committed:
Don't need lock/safe-point for ShenandoahOldGeneration::expend_promoted

pengxiaolong committed:
Fix alignment issue with plab alloc

pengxiaolong committed:
Always use snapshot of top when cacluclate aligned size for lab alloc

pengxiaolong committed:
Revert the changes to use ShenandoahOldCollectorAllocator, not ready yet

pengxiaolong committed:
More refectors for ShenandoahAllocator

pengxiaolong committed:
Add missing retire_from_partition call

pengxiaolong committed:
fix

pengxiaolong committed:
Only try to reserve alloc regions when there is at least one refreshable alloc region

pengxiaolong committed:
Fix valid logic checking if a region is ready for retire after atomic allocation

pengxiaolong committed:
_ready_for_retire of ShenandoahHeapRegionReadyForRetireChecker should be a ref

pengxiaolong committed:
Move two internal methods of ShenandoahFreeSet to private

pengxiaolong committed:
1. Always use AtomicAccess::load; 2. ShenandoahOldCollectorAllocator always delegate to  _free_set->allocate_for_collector

pengxiaolong committed:
ShenandoahOldCollectorAllocator should not take heap lock, but assert heap locked.

pengxiaolong committed:
Reserve region a alloc region if capacity >= PLAB::min_size

pengxiaolong committed:
Remove unused code after refector

pengxiaolong committed:
Add flags: ShenandoahCollectorAllocRegions, ShenandoahMutatorAllocRegions

pengxiaolong committed:
Set _alloc_start_index to 0 if _alloc_region_count is 1

pengxiaolong committed:
Yield to safepoint if needed

pengxiaolong committed:
Move ShenandoahAllocator::yield_to_safepoint to cpp

pengxiaolong committed:
inline atomic_allocate_in

pengxiaolong committed:
Fix a bug in refresh_alloc_regions

pengxiaolong committed:
Exclude regions if affiliation doesn't match

pengxiaolong committed:
Remove _direct_alloc_mutators from ShenandoahHeapRegion

pengxiaolong committed:
Fix issue in refresh_alloc_regions

pengxiaolong committed:
Fix and improvements after debugging

pengxiaolong committed:
update find_heap_region_for_allocation_internal and reserve_alloc_regions_internal

pengxiaolong committed:
Use cmpxchg to update to watermark because we may update it concurrently from multiple threads after allocations

pengxiaolong committed:
Revert changes for watermark

pengxiaolong committed:
Add debug logs for ShenandoahAllocator

pengxiaolong committed:
Ensure all alloc regions are released before choose cset/rebuild free-set

pengxiaolong committed:
Add debug log for alloc result

pengxiaolong committed:
Use concurrent_set_update_watermark in ShenandoahAllocator

pengxiaolong committed:
Add assert !region->is_active_alloc_region() assert when choose_collection_set

pengxiaolong committed:
Add attempt_allocation_slow

pengxiaolong committed:
Remove dead code from FreeSet

pengxiaolong committed:
touch up

pengxiaolong committed:
Reorder slow-path to solve unexpected crash

pengxiaolong committed:
Release all release_alloc_regions even before ShenandoahVerify in ShenandoahOldGC::op_final_mark

pengxiaolong committed:
Move all the old alloc logic in ShenandoahHeap to ShenandoahOldCollectorAllocator

pengxiaolong committed:
Attempt to fix assert "Cannot remove more regions than are present"

pengxiaolong committed:
Remove ShenandoahFreeSet::allocate_for_mutator

pengxiaolong committed:
Fix corner cases in which heap accountings are not updated.

pengxiaolong committed:
increase region counts of partition when unretire a region

pengxiaolong committed:
Move inline function definition to hpp file

pengxiaolong committed:
Try to satisfy alloc req before install alloc region

pengxiaolong committed:
Fix build error in some build variants

pengxiaolong committed:
Add #include "memory/allocation.hpp"

pengxiaolong committed:
Comment out sanity check in recompute_total_affiliated

pengxiaolong committed:
Add missing include for "memory/padded.hpp"

pengxiaolong committed:
Add missing include for memory/padded.inline.hpp

pengxiaolong committed:
Fix bug in attempt_allocation

pengxiaolong committed:
Always reserve affiliated regions first for collector allocator

pengxiaolong committed:
Wrong short-cut condition

pengxiaolong committed:
Fix bug

pengxiaolong committed:
Bug fix

pengxiaolong committed:
Bug fix!

pengxiaolong committed:
Optimize

pengxiaolong committed:
More logs

pengxiaolong committed:
Update recompute total used and total affiliated more often

pengxiaolong committed:
offset-by-one issue in new impl of reserve_alloc_regions_internal

pengxiaolong committed:
Create ShenandoahHeapAccountingUpdater after heap lock

pengxiaolong committed:
Disable sanity in recompute_total_affiliated

pengxiaolong committed:
Don't promote active alloc region

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-refactor

pengxiaolong committed:
defer accounting recomputation is not handled properly (plus some code formating)

pengxiaolong committed:
Reserve alloc regions before the clients(mutator/collector) needs to use them

pengxiaolong committed:
Sort header includes

pengxiaolong committed:
Fix head sort issue

pengxiaolong committed:
Fix ident

pengxiaolong committed:
Rename fields of struct ShenandoahAllocRegion to remove prefix `_`

pengxiaolong committed:
Rewrite attempt_allocation_in_alloc_regions and avoid using AtomicAccess::load_acquire

pengxiaolong committed:
memory order relaxed when adjust_alloc_metadata

pengxiaolong committed:
Simplify reset_age

pengxiaolong committed:
Remove AtomicAccess when access affiliation

pengxiaolong committed:
Revert ShenandoahOldGeneration

pengxiaolong committed:
Remove use of load_acquire and release_store

pengxiaolong committed:
Unset active_alloc_region flag when the region is out of space

pengxiaolong committed:
Use AtomicAccess::store instead of AtomicAccess::release_store

pengxiaolong committed:
Set default value of ShenandoahMutatorAllocRegions to 8

pengxiaolong committed:
Remove unnecessary assert

pengxiaolong committed:
Minor fix

pengxiaolong committed:
For Mutator only reserve regions as active alloc region if the free space is >= PLAB::max_size()

pengxiaolong committed:
Fix build

pengxiaolong committed:
Check if partition is empty before diving into find_heap_region_for_allocation_internal

pengxiaolong committed:
Not try steal region from mutator in reserve_alloc_regions

pengxiaolong committed:
Use template for in the impl of reserve_alloc_regions and find_heap_region_for_allocation

pengxiaolong committed:
MAX_ALLOC_REGION_COUNT should be 128

pengxiaolong committed:
Remove unnecessary change and tidy up

pengxiaolong committed:
Remove junk code

pengxiaolong committed:
Merge remote-tracking branch 'origin/master' into cas-alloc-1

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

pengxiaolong committed:
Port the fix of JDK-8372566

pengxiaolong committed:
Make ShenandoahAllocator as template class to make compiled code more efficient for each alloc partition

pengxiaolong committed:
Remove unnecessary `enum class ShenandoahFreeSetPartitionId : uint8_t` in shenandoahAllocator.php

pengxiaolong committed:
Fix a typo

pengxiaolong committed:
Declare ShenandoahFreeSetPartitionId as enum instead of enum class

pengxiaolong committed:
Add missing header for ShenandoahFreeSetPartitionId

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,466 @@
+                                                /* UnaffiliatedChangesAreYoungNeutral */ false>();
+          break;
+        case ShenandoahFreeSetPartitionId::OldCollector:
+          _free_set->recompute_total_used</* UsedByMutatorChanged */ true,
```
These parameters seem overly conservative.  Can we distinguish what needs to be recomputed?
Normally, OldCollector allocation does not change UsedByMutator or UsedByCollector.  It will only change MutatorEmpties if we did flip_to_old.  It will normally not changed OldCollectorEmpties (unless it flips multiple mutator to OldCollector.  if might flip one region from mutator, but that region will not be empty after we allocate fro it...

pengxiaolong replied:
We don't really know what need to be recompute until the allocation finishes, we can make it less  conservative, but then we needs more code branches here because the template methods require explicit template parameters.

I'll create to ticket to follow up on this, given that I also want to see if we can defer the recomputation to the read side, if we can do that we don't even need the ShenandoahHeapAccountingUpdater here.

pengxiaolong replied:
I have created a bug for the improvement of accounting update: https://bugs.openjdk.org/browse/JDK-8373371

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,466 @@
+  // The region returned by find_heap_region_for_allocation must have sufficient free space for the allocation it if it is not nullptr
+  if (r != nullptr) {
+    bool ready_for_retire = false;
+    obj = atomic_allocate_in(r, false, req, in_new_region, ready_for_retire);
```
Not sure why we use atomic_allocate_in() here.  We hold the heap lock so we don't need to use atomic operations.  
We should clarify with comments.

pengxiaolong replied:
It is not really necessary to `atomic_allocate_in` here, but I wanted reuse some of the codes in atomic_allocate_in, we can discuss this later, I can change it back to non-atomic version.

kdnilsen replied:
Would prefer not to use the atomic_allocate code here.  If you want to reuse code, maybe you can refactor allocate_in with an <ATOMIC> template argument.

I notice that this PR makes lots of ShenandoahHeapRegion variables volatile: _age, _youth, _top, _tlab_allocs, _gclab_allocs, _plab_allocs.  That will cause less efficient code to be generated whenever we are accessing this data from "behind" the GC heap lock, which should be rare, I guess.

I raised some concerns/issues about the race that happens when we move a region from the directly-allocatable set into the global-heap-lock-protected ShenandoahFreeSet partition.

Here's the scenario that I'm concerned about:

1. A mutator obtains pointer to directly allocatable region R
2. A second mutator performs a refresh, moving region R out of directly allocatable set (for whatever reason)
3. Region R is now eligible to satisfy allocations from behind the global heap lock
4. Some third mutator thread acquires the heap lock and fetches top for region $
5. The first mutator performs its allocation within the same region R, not recognizing a CAS conflict
6. This third mutator allocates from region R at top, without using CAS.  So both mutators think they own the same object

I think this is not a problem if the "only" reason at step 2 above that we move region R out of directly allocatable set is because R is ready to be retired.  In that case, there will be no subsequent heap-locked allocations in regions R.  However, I anticipate the day in not-too-distant future when we will want to refresh regions even when they are not ready to be retired.  Specifically, as we move rebuild-freeset out of safepoints, we will want to refresh regions before we acquire heap-lock to do rebuild, with the goal of making sure there is sufficient directly allocatable memory available that no mutator will be stalled because it needs to allocate during the time that the heap remains locked for the rebuild operation.

So I suppose that if we always use atomic_allocate() even for allocations that happen while holding the heap lock, we won't have this problem.  If we decide to keep this architecture, there should be comments explaining why we are doing it this way.  (I am not real happy that we have to "pay the cost" of CAS in addition to paying the cost of global heap lock, but I think these allocations should be very rare.  It seems this would only come up if, for example, a mutator wanted to allocate an object that is 1/2 the heap region size, and none of the directly allocatable regions have that much available memory.)

My original proposal was to have a volatile_top which is used by CAS allocation and a nonvolatile_top that is used by heap-lock allocation.  When we make a region directly allocatable, we copy its nonvolatile_top to the volatile_top.  When we take a directly allocatable region and move it into the heap-locked freeset, we use CAS to set its volatile_top to end before we place the region into the freeset partition, assigning to nonvolatile_top the value held in volatile_top before the CAS operation.

Whatever solution is used for this needs to be documented in the code.  Feel free to copy and paste from this github comment.

kdnilsen reviewed (commented):
I'm still reading through the code, but have these comments far...

pengxiaolong commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,466 @@
+template <ShenandoahFreeSetPartitionId ALLOC_PARTITION>
+HeapWord* ShenandoahAllocator<ALLOC_PARTITION>::attempt_allocation(ShenandoahAllocRequest& req, bool& in_new_region) {
+  if (_alloc_region_count == 0u) {
+    ShenandoahHeapLocker locker(ShenandoahHeap::heap()->lock(), _yield_to_safepoint);
```
I'll add comments on this, _alloc_region_count == 0 means we don't want to use any shared alloc region, it will also allocate with a heap lock, ideally the performance should be same as before, so it always simply find a region with enough space and allocate in the region.

pengxiaolong replied:
I have added comments on those functions, I'll keep adding more for those missing comments; meanwhile I am trying to avoid excessive comment, pleas point out if any of the comments is not clear.

pengxiaolong replied:
I have removed the branch handling  _alloc_region_count == 0, also updated comments.

pengxiaolong commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,466 @@
+
+template <ShenandoahFreeSetPartitionId ALLOC_PARTITION>
+HeapWord* ShenandoahAllocator<ALLOC_PARTITION>::attempt_allocation_slow(ShenandoahAllocRequest& req, bool& in_new_region) {
+  ShenandoahHeapLocker locker(ShenandoahHeap::heap()->lock(), _yield_to_safepoint);
```
It is not an error, before calling into attempt_allocation_slow, it already called attempt_allocation_in_alloc_regions once and failed to allocate, slow path is always with heap lock.

After taking the lock, we should try the attempt_allocation_in_alloc_regions right away, because other mutator thread may have refreshed the alloc regions while holding the lock.

pengxiaolong replied:
The concern is the not "this count is getting incremented "during" our allocation", it is the the case when it get decremented because other mutators may have already refreshed all alloc regions before current mutator getting heap lock, we have call attempt_allocation_in_alloc_regions again after successfully acquiring heap lock because of this.

Same design can be also found in G1, Parallel and Serial GC's CAS allocator.

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong committed:
Address PR comments

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

pengxiaolong committed:
Expend promoted from ShenandoahOldCollectorAllocator

pengxiaolong committed:
Fix build failure after merge

pengxiaolong committed:
Some comments updates as suggested in PR review

pengxiaolong commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,466 @@
+  HeapWord* obj;
+  size_t min_free_words = req.is_lab_alloc() ? req.min_size() : req.size();
+  ShenandoahHeapRegion* r = _free_set->find_heap_region_for_allocation(ALLOC_PARTITION, min_free_words, req.is_lab_alloc(), in_new_region);
+  // The region returned by find_heap_region_for_allocation must have sufficient free space for the allocation it if it is not nullptr
```
fixed.

pengxiaolong commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,478 @@
+ShenandoahAllocator<ALLOC_PARTITION>::ShenandoahAllocator(uint const alloc_region_count, ShenandoahFreeSet* free_set):
+  _alloc_region_count(alloc_region_count), _free_set(free_set), _alloc_partition_name(ShenandoahRegionPartitions::partition_name(ALLOC_PARTITION)) {
+  if (alloc_region_count > 0) {
+    _alloc_regions = PaddedArray<ShenandoahAllocRegion, mtGC>::create_unfreeable(alloc_region_count);
```
Rethinking about the the PaddedArray used here, we may not really need it.
Allocator has multiple shared alloc regions for CAS, and only refreshes them when all of them run out of usable memory, so _alloc_regions won't be frequently updated, the PaddedArray here should have a negative performance impact.

kdnilsen replied:
Are you running any experiments (on different hardware configurations) to test your assumptions about this?

pengxiaolong replied:
Here is specjbb results I got, heap size is 8G, 8 cores:


|                     | PaddedArray    | Raw array |         |
| ------------------- | -------------- | --------- | ------- |
| 1st                 | Max jOPS       | 17893     | 18092   | 1.11%   |
| Critical jOPS       | 15490          | 15765     | 1.78%   |
| 2nd                 | Max jOPS       | 18219     | 18092   | \-0.70% |
| Critical jOPS       | 15625          | 15498     | \-0.81% |
| Average             | Max jOPS       | 18056     | 18092   | 0.20%   |
| Critical jOPS       | 15557.5        | 15631.5   | 0.48%   |
| Standard deviation  | Max jOPS Stdev | 163       | 0       |         |
| Critical jOPS Stdev | 67.5           | 133.5     |         |

I didn't see huge difference, using raw array tends to be slightly better.

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

pengxiaolong committed:
Fix build error after merging from tip

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.hpp:
```diff
@@ -0,0 +1,132 @@
+  // Refresh new alloc regions, allocate the object in the new alloc region.
+  int refresh_alloc_regions(ShenandoahAllocRequest* req = nullptr, bool* in_new_region = nullptr, HeapWord** obj = nullptr);
+#ifdef ASSERT
+  virtual void verify(ShenandoahAllocRequest& req) { }
```
Need a comment to explain what verify does.  Is this simply checking to make sure the req is "properly formatted"?

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,478 @@
+
+  for (uint i = 0; i < _alloc_region_count; i++) {
+    ShenandoahAllocRegion& alloc_region = _alloc_regions[i];
+    ShenandoahHeapRegion* r = AtomicAccess::load(&alloc_region.address);
```
We've got heap lock and at safepoint.  Do not need AtomicAccess here.  That is more costly than necessary.  I prefer to use regular fetch.  If you prefer to keep AtomicAccess, please provide a comment in the code explaining why and we will revist.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,478 @@
+      if (r->is_active_alloc_region()) {
+        r->unset_active_alloc_region();
+      }
+      AtomicAccess::store(&alloc_region.address, static_cast<ShenandoahHeapRegion*>(nullptr));
```
Same here.  We do not need AtomicAccess.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,478 @@
+  }
+  _free_set->partitions()->decrease_used(ALLOC_PARTITION, total_free_bytes);
+  _free_set->partitions()->increase_region_counts(ALLOC_PARTITION, total_regions_to_unretire);
+  accounting_updater._need_update = true;
```
Here is where you know which tallies have been affected by this operation.  This is where you should specialize the calls to freeset recompute_total_used() and recompute_total_affiliated().  Either call those from here, or add parameters to your accounting_updater object so that you do not have to overcompute each operation.

pengxiaolong replied:
This is not the only place, there are many other places in the allocation code path as well, I can specialize the calls to  recompute_total_used and recompute_total_affiliated in the places, but that will make the code necessary rambling, I'd rather to solve this issue and simplify it in https://bugs.openjdk.org/browse/JDK-8373371.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.hpp:
```diff
@@ -0,0 +1,132 @@
+  HeapWord* attempt_allocation_from_free_set(ShenandoahAllocRequest& req, bool& in_new_region);
+
+  // Attempt to allocate in a shared alloc region using atomic operation without holding the heap lock.
+  // Returns nullptr and overwrites regions_ready_for_refresh with the number of shared alloc regions that are ready
```
Suggest this edit:
// Overwrites regions_ready_for_refresh with a lower bound on the number of shared alloc regions that are ready
// to be retired during execution of this "do_fast_allocation" function.  Returns nullptr if the allocation request could
// not be fulfilled after a single traversal of directly allocatable regions.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.inline.hpp:
```diff
@@ -125,15 +126,117 @@ HeapWord* ShenandoahHeapRegion::allocate(size_t size, const ShenandoahAllocReque
+  }
+};
+
+HeapWord* ShenandoahHeapRegion::allocate_atomic(size_t size, const ShenandoahAllocRequest& req, bool &ready_for_retire) {
```
Suggest we add a fourth arg: int &contended
We initialize contended to zero

pengxiaolong replied:
What do we want to track with the `contended` arg? If we want to track the number of threads allocating simultaneously from the same region or from same shared alloc region index, we can't archive either of them with the arg `&contended `.

Let's discuss this later, I need to understand the intention/idea before making this change.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.inline.hpp:
```diff
@@ -125,15 +126,117 @@ HeapWord* ShenandoahHeapRegion::allocate(size_t size, const ShenandoahAllocReque
+      retire_checker._remnant_free_words = free_words;
+      return nullptr;
+    }
+  }
```
Before iterating, increment contended by 1

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.inline.hpp:
```diff
@@ -125,15 +126,117 @@ HeapWord* ShenandoahHeapRegion::allocate(size_t size, const ShenandoahAllocReque
+  }
+}
+
+HeapWord* ShenandoahHeapRegion::allocate_lab_atomic(const ShenandoahAllocRequest& req, size_t &actual_size, bool &ready_for_retire) {
```
Suggest we add a fourth arg: int &contended
We initialize contended to zero

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,478 @@
+      // For GC allocations, we advance update_watermark because the objects relocated into this memory during
+      // evacuation are not updated during evacuation.  For both young and old regions r, it is essential that all
+      // PLABs be made parsable at the end of evacuation.  This is enabled by retiring all plabs at end of evacuation.
+      region->concurrent_set_update_watermark(region->top());
```
There's a race here.  Multiple mutators may be updating watermark in parallel.  It may be that the mutator who most recently allocated is not the mutator who makes the "most recent" overwrite of set_update_watermark().

I think the better fix is to remove this code.  Update refs should just assume that update watermark equals top for any region in the Old gen, and for any region that was in the Collector partition.  Maybe we use a Sentinel value for update_watermark on all such regions.  Just overwrite update_watermark(nullptr)?  And check for this in update-refs?  Needs a solution, and solution needs to be documented.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,478 @@
+  assert(regions_ready_for_refresh == 0u && in_new_region == false && alloc_start_index < _alloc_region_count, "Sanity check");
+  uint i = alloc_start_index;
+  do {
+    if (ShenandoahHeapRegion* r =  nullptr; (r = _alloc_regions[i].address) != nullptr && r->is_active_alloc_region()) {
```
Note that there is a race (and performance overhead) with checking r->is_active_alloc_region().  Though a region might be active when we check it here, it may be inactive by the time we attempt to atomic_allocate_in().

This is one reason I prefer to use "volatile_top == end" to denote !is_active_alloc_region.  This way, you only have to check once (rather than checking is_active() and then checking has_available()).  And there is no race between when you check and when you attempt to allocate.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,478 @@
+  if (refreshable_alloc_regions > 0) {
+    // Step 2: allocate region from FreeSets to fill the alloc regions or satisfy the alloc request.
+    ShenandoahHeapRegion* reserved[MAX_ALLOC_REGION_COUNT];
+    int reserved_regions = _free_set->reserve_alloc_regions(ALLOC_PARTITION, refreshable_alloc_regions,
```
I request we get rid of the min_free_words argument to free_set->reserve_alloc_regions().  See comments in the called function.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,478 @@
+        }
+        log_debug(gc, alloc)("%sAllocator: Storing heap region %li to alloc region %i",
+          _alloc_partition_name, reserved[i]->index(), refreshable[i]->alloc_region_index);
+        AtomicAccess::store(&refreshable[i]->address, reserved[i]);
```
Should not need to perform AtomicAccess because we hold the heap lock here.

pengxiaolong replied:
When we store a region to alloc region array, AtomicAccess::store will ensure other threads can see it immediately, but the read after holding heap lock has been updated to not AtomicAccess::load.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,478 @@
+template <ShenandoahFreeSetPartitionId ALLOC_PARTITION>
+HeapWord* ShenandoahAllocator<ALLOC_PARTITION>::allocate(ShenandoahAllocRequest &req, bool &in_new_region) {
+#ifdef ASSERT
+  verify(req);
```
Insert a comment above verify(): "Conform that req corresponds to ALLOC_PARTITION"

pengxiaolong replied:
Done, thanks.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,478 @@
+  ShenandoahAllocator(0u, free_set) {
+  _yield_to_safepoint = false;
+}
+
```
I suppose ShenandoahCollectorAllocator::randomize_start_index() might be a no-op.  On the other hand, it would probably be better to use a random index for ShenandoahCollectorAllocator as well.  We don't want to hobble one GC worker more than the others just because its preferred start index happens to hold a retire-ready region.

pengxiaolong replied:
The code you put comments on is ShenandoahOldCollectorAllocator, which delegate the allocation to ShenandoahFreeSet in the PR, 0u in passed to ShenandoahAllocator constructor is the number of shared alloc regions, which should be 0. I'll work on ShenandoahOldCollectorAllocator to make it support CAS as well. 

The start index of ShenandoahCollectorAllocator is slightly different from ShenandoahMutatorAllocator, for GC worker we can always use `worker_id % _alloc_region_count`, but there is some corner cases we need to consider, e.g.  Java thread may call into ShenandoahCollectorAllocator from load barrier code. Currently for non worker threads, the start index is always 0, I think this is the part we can change it to random.

pengxiaolong replied:
I have created the ticket, here is it https://bugs.openjdk.org/browse/JDK-8381331

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,478 @@
+  return Thread::current()->is_Worker_thread() ? WorkerThread::worker_id() % _alloc_region_count : 0u;
+}
+
+HeapWord* ShenandoahOldCollectorAllocator::allocate(ShenandoahAllocRequest& req, bool& in_new_region) {
```
Confer with William Kemper about this.  He is working on a change that may simplify the handling of PLABs, in which case ShenandoahOldCollectorAllocator can behave the same as ShenandoahCollector.

kdnilsen replied:
Alternatively, I don't think it would be too terribly difficult to implement try_allocate_aligned() function to support fast (CAS) allocation of aligned PLABs.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.inline.hpp:
```diff
@@ -198,6 +301,16 @@ inline void ShenandoahHeapRegion::set_update_watermark(HeapWord* w) {
   AtomicAccess::release_store(&_update_watermark, w);
 }
 
+inline void ShenandoahHeapRegion::concurrent_set_update_watermark(HeapWord* w) {
```
See comment elsewhere in my feedback.  I think we may want to use a special sentinel value to denote that watermark for Collector and OldCollector regions.  For both of these, there is essentially not watermark value.  If we try to set the value to top() from within a CAS-allocating mutator thread, we can end up setting watermark to the not-most-recent value of top(), which would result in misbehavior during update refs.

kdnilsen reviewed (changes requested):
This is a huge PR.  Thanks for working through all the details to get this working.  I've identified several issues that I believe require some further attention.  We can discuss in a meeting if that would be helpful.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,478 @@
+  // Fast path: start the attempt to allocate in alloc regions right away
+  HeapWord* obj = attempt_allocation_in_alloc_regions(req, in_new_region, alloc_start_index(), dummy);
+  if (obj != nullptr) {
+    return obj;
```
After further thought, am thinking the threshold for refresh_alloc_regions() might be if (regions_ready_for_refresh >= _alloc_region_count / 2).  That would reduce the number of slow paths through the allocator.  If we can re-randomize the thread-local start indexes when their original start index hits a retire-able region, this might work ok.

pengxiaolong replied:
I have added support for this, it works as:

```
obj = fast-path-alloc();
if (obj != nullptr && regions_ready_for_refresh < _alloc_region_count / 2) {
  return obj;
}
if (obj = nullptr) {
   obj = slow-path-alloc();
} else {
   ShenandoahHeapLocker locker(ShenandoahHeap::heap()->lock(), false);
   refresh_alloc_regions();
}
```

If fast-path-alloc succeeds but determines that there are more than 50% of alloc regions are ready for retiring, when it take heap heap lock to refresh alloc regions, it CANNOT yield to safepoint, because the thread is holding uninitialized obj.

pengxiaolong committed:
Only use atomic allocation when allocate from shared alloc regions

pengxiaolong committed:
Update assert message

pengxiaolong committed:
Update code comments

pengxiaolong committed:
No need to use Atomic::load to read shared alloc region in refresh_alloc_regions

pengxiaolong committed:
No need to use Atomic::load to read shared alloc region in release_alloc_regions

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong committed:
Invalid assert

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong committed:
Avoid AtomicAccess::load in attempt_allocation_in_alloc_regions when called from attempt_allocation_slow

pengxiaolong committed:
Use template parameter for ShenandoahAllocator::verify

pengxiaolong committed:
Merge branch 'cas-alloc-1' of https://github.com/pengxiaolong/jdk into cas-alloc-1

pengxiaolong committed:
add include header shenandoahAllocRequest.hpp

pengxiaolong committed:
Fix header file order

pengxiaolong committed:
Fix include error

pengxiaolong committed:
1. Allow to reserve alloc region with free space >= PLAB::min_size()
2. Other minor fixes

pengxiaolong committed:
Add _epoch_id to ShenandoahAllocator to trace the update of shared alloc regions

pengxiaolong committed:
Remove necessary atomic load

pengxiaolong committed:
Add comments for public methods

pengxiaolong committed:
Move thread locals to ShenandoahThreadLocalData

pengxiaolong committed:
Make release_alloc_regions and reserve_alloc_regions non-virtual methods

pengxiaolong committed:
UUpdate the comments on method allocate

pengxiaolong committed:
Update comments for function alloc_start_index

pengxiaolong commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.cpp:
```diff
@@ -0,0 +1,478 @@
+
+  ShenandoahHeapAccountingUpdater accounting_updater(_free_set, ALLOC_PARTITION);
+
+  if (regions_ready_for_refresh > 0u) {
```
To actually track the accurate number of `refresh_regions_available`, we need to add three counters: one for each partition. Since we always retire the region is the free is less than PLAB::min_size(), we may already have it in free set partitions(_region_counts?).

There is also complexity here, for Collector and OldCollector, we allow them to transfer regions from Mutator partition,  we have to consider FREE regions in mutator partition to calculate  "refresh_regions_available" for Collector/OldCollector.

It should be rare case when the entire heap is almost filled up, I am wondering how much benefit we could get from doing that, if you are ok with it I'd suggest to consider this optimization later.

pengxiaolong committed:
Eagerly refresh alloc region if there are 1/2 or more of alloc regions ready for retire

pengxiaolong committed:
Add virtual back for release_alloc_regions and reserve_alloc_regions to fix link error

pengxiaolong committed:
Remove the support of 0 for flags ShenandoahMutatorAllocRegions and ShenandoahCollectorAllocRegions

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong committed:
Fix assert when after eagerly refresh alloc regions after fast allocation

pengxiaolong committed:
While eagerly refresh alloc regions, thread should not yield to safepoint because it is holding uninitialized new object

pengxiaolong committed:
More eagerly to refresh alloc regions in attempt_allocation_slow since it is holding heap lock

pengxiaolong committed:
typo

pengxiaolong committed:
Code format

pengxiaolong committed:
More accurate census noise

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

pengxiaolong committed:
Use simple array instead of PaddedArray to store alloc regions

pengxiaolong committed:
Update code as suggested during the review meeting

pengxiaolong committed:
Only to set update_watermark when allocate from collector/old collector partition

pengxiaolong committed:
Add _volatile_top for atomic allocation in heap region w/o heap lock to address potential race condition when refresh alloc regions

pengxiaolong committed:
Code format

pengxiaolong committed:
Remove refreshed_regions

pengxiaolong committed:
Set _volatile_top to nullptr(instead of end of region) for region that is not active for CAS alloc

pengxiaolong committed:
Fix assert in recycle_internal

pengxiaolong committed:
Fix assert in concurrent_set_update_watermark

pengxiaolong committed:
Load _volatile_top once in free_bytes_for_atomic_alloc

pengxiaolong committed:
Fix new race condition in unset_active_alloc_region which sync _atomic_top back to _top, other threads may see stale _top if not holding heap lock

pengxiaolong committed:
Add more code comments

pengxiaolong committed:
Enforce memory ordering for read/write of _atomic_top to fix crash on MacOS

pengxiaolong committed:
Enforce memory order by adding fence

pengxiaolong committed:
Remove one redundant fence

pengxiaolong committed:
More comments

pengxiaolong committed:
Merge branch 'cas-alloc-1' of https://github.com/pengxiaolong/jdk into cas-alloc-1

pengxiaolong committed:
Remove _active_alloc_region flag and directly use field _atomic_top to determine atomic alloc region

pengxiaolong committed:
Rename function is_active_alloc_region to is_atomic_alloc_region

pengxiaolong committed:
Missed place calling function is_active_alloc_region

pengxiaolong committed:
Remove use AtomicAccess for _top if permits

pengxiaolong committed:
Invoke assert_bounds after updating accounting from ShenandoahAllocator

pengxiaolong committed:
Not update accounting if no region is unretired to partition

pengxiaolong committed:
FullGC should not update accounting when release alloc regions

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

pengxiaolong committed:
Fix build error after merging tip

pengxiaolong committed:
Fix crash after merging from tip(due to the new free-set rebuild calls)

pengxiaolong committed:
Update ProblemList to enable the tests affected by the change for accounting update

pengxiaolong committed:
Add assert to functions new_top & set_new_top; fix typo

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

pengxiaolong committed:
Proactively steal region from mutator view when refresh alloc regions for collector/old collector

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

pengxiaolong committed:
Fix build failure merging tip

pengxiaolong committed:
Lazily replenish alloc regions instead of proactively doing in alloc code path

pengxiaolong committed:
Convert AtomicAccess to Atomic<T> for ShenandoahAllocator

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

pengxiaolong committed:
Minor fix/typos suggested by kiro

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

pengxiaolong committed:
Count the remaining bytes of mutator allocator as not used for more accurate accounting

pengxiaolong committed:
More accurate method/variable names

pengxiaolong committed:
Bug fix

pengxiaolong committed:
More code improvements

pengxiaolong committed:
Remove duplicate semi-colon

pengxiaolong committed:
Fix Copyright format issue

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong committed:
Add padding around _atomic_top

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.hpp:
```diff
@@ -0,0 +1,180 @@
+  // Collector calls this in preparation for choosing a collection set and/or rebuilding the freeset.
+  virtual void release_alloc_regions(bool should_update_accounting = true);
+
+  // Caller must hold the heap lock at safepoint. This causes us to set aside N regions as directly allocatable
```
Can comment clarify what is N?

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahAllocator.hpp:
```diff
@@ -0,0 +1,180 @@
+  ShenandoahCollectorAllocator(ShenandoahFreeSet* free_set);
+};
+
+// Currently ShenandoahOldCollectorAllocator delegate allocation handling to ShenandoahFreeSet,
```
Minor: s/delegate/delegates/

pengxiaolong committed:
Address PR comments

pengxiaolong committed:
Avoiding calling reset_age and set_update_watermark in fast alloc path

pengxiaolong committed:
Remove concurrent_set_update_watermark from header file

pengxiaolong committed:
Convert _tlab_allocs/_gclab_allocs/_plab_allocs to Atomic<size_t>

pengxiaolong committed:
Fix build errors

pengxiaolong committed:
Sort includes

pengxiaolong committed:
Revert the change making get_usable_free_words static

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong committed:
Remove unnecessary virtual modifier

pengxiaolong committed:
Remove use_affiliated_first logic from collector allocation

pengxiaolong committed:
Remove unnecessary loop

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

pengxiaolong committed:
nit: rename functions

pengxiaolong committed:
nit: comments

pengxiaolong commented on src/hotspot/share/gc/shared/plab.cpp:
```diff
@@ -37,6 +37,10 @@ size_t PLAB::min_size() {
   return align_object_size(MAX2(MinTLABSize / HeapWordSize, (size_t)oopDesc::header_size())) + CollectedHeap::lab_alignment_reserve();
 }
 
+size_t PLAB::min_size_bytes() {
```
Thanks, I have updated the comment in the header file for all these methods, also renamed the method name as `xx_byte_size` is already used in existing common code.

pengxiaolong committed:
nit: comments

pengxiaolong commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:
```diff
@@ -2313,17 +2267,18 @@ void ShenandoahHeap::reset_bytes_allocated_since_gc_start() {
   }
-  ShenandoahHeapLocker locker(lock());
-  _free_set->reset_bytes_allocated_since_gc_start(unaccounted_bytes);
+  _free_set->reset_bytes_allocated_since_gc_start(unaccounted_bytes + mutator_allocator_remaining_bytes);
```
The part actually confused me while I made the change, if I apply the change as your suggestion, we will see underflow issue resulting in super high allocation rate(wrong sampling) in multiple tests which leads to excessive GC cycles.  Let's talk about this tomorrow.

pengxiaolong committed:
nit: comments to explain why mutator_allocator_remaining_bytes needs to be added back when reset bytes_allocated_since_gc_start

pengxiaolong committed:
Use load_acquire when read _epoch_id

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

pengxiaolong commented:
Attach specjbb2015 from our performance pipeline, the improvement in specjbb2015 is very stable in recnetly pipeline runs after the changes to optimize water mark and region age update in CAS allocator:

<img width="1204" height="417" alt="Specjbb2015-128g" src="https://github.com/user-attachments/assets/4e684612-8ec9-4f6b-9d3e-2f53648158ad" />

<img width="1182" height="687" alt="Specjbb2015" src="https://github.com/user-attachments/assets/3d138e6f-fe41-4f68-863b-780f687fb6cc" />

pengxiaolong committed:
Fix memory ordering, UB, and CAS contention in CAS allocator

pengxiaolong committed:
Merge branch 'master' into cas-alloc-1

pengxiaolong committed:
Split top() into concurrent default and stable_top() for CAS alloc regions

pengxiaolong committed:
Use Atomic<T> for _age and _atomic_top fields

pengxiaolong committed:
Restore atomic access to _affiliations with release/acquire ordering

pengxiaolong committed:
Fix format specifiers, typos, and minor style issues

pengxiaolong committed:
Assert heap lock in remaining_bytes, drop redundant reset_age in CAS alloc, expose _atomic_top to SA

pengxiaolong committed:
Add jtreg tests for CAS allocator flag combinations and contention

pengxiaolong committed:
fix: sync _top once on CAS success and reset_age after the loop in unset_active_alloc_region

pengxiaolong committed:
fix: sync _top before CAS in unset_active_alloc_region

pengxiaolong committed:
fix: new jtreg tests miss -XX:+UnlockDiagnosticVMOptions

pengxiaolong committed:
Reserve mutator alloc regions after abbreviated degenerated GC

pengxiaolong committed:
Comment: pair reserve-time inflation with read-time compensation

Cross-reference ShenandoahFreeSet::reserve_alloc_regions_internal at
every 'bytes_allocated - mutator_allocator_remaining' subtraction site.
No behavior change.

pengxiaolong committed:
Include mutator allocator remaining bytes in unsafe_max_tlab_alloc

pengxiaolong committed:
Tame TestCASAllocContention: fewer threads, shorter run, smaller retention

pengxiaolong committed:
Document _collector_allocator_reserved ordering contract

pengxiaolong committed:
Merge branch 'openjdk:master' into cas-alloc-1

shipilev commented on src/hotspot/share/gc/shared/plab.cpp:
```diff
@@ -37,10 +37,18 @@ size_t PLAB::min_size() {
   return ThreadLocalAllocBuffer::max_size();
 }
 
+size_t PLAB::max_byte_size() {
```
I don't think this is used anywhere.

shipilev commented on src/hotspot/share/gc/shared/plab.cpp:
```diff
@@ -37,10 +37,18 @@ size_t PLAB::min_size() {
   return align_object_size(MAX2(MinTLABSize / HeapWordSize, (size_t)oopDesc::header_size())) + CollectedHeap::lab_alignment_reserve();
 }
 
+size_t PLAB::min_byte_size() {
```
There is a single use. Inline the implementation to avoid extra shared code changes.

pengxiaolong replied:
Yes, It is single use for now, I will remove it to avoid change in shared code.

shipilev commented on src/hotspot/share/gc/shenandoah/shenandoahAffiliation.hpp:
```diff
@@ -25,6 +25,8 @@
 #ifndef SHARE_GC_SHENANDOAH_SHENANDOAHAFFILIATION_HPP
 #define SHARE_GC_SHENANDOAH_SHENANDOAHAFFILIATION_HPP
 
+#include "utilities/debug.hpp"
```
Do you need it? If you don't use anything from the header in this file, move the include where it is actually needed.

pengxiaolong replied:
I will fix this, there is no change in this file, I don't think I need the include for utilities/debug.hpp, I may reverted this changes in file but forgot to remove the header.

shipilev commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp:
```diff
@@ -410,16 +411,31 @@ static double saturate(double value, double min, double max) {
+  size_t allocated = 0;
+  size_t allocated_bytes_since_last_sample = 0;
+
+  {
```
I think this clashes with https://github.com/openjdk/jdk/pull/31047, which simplifies this path.

pengxiaolong replied:
Yes, I saw William's PR.  If #31049 is merged before this PR, I'll rework on this part.

shipilev reviewed (changes requested):
No, this requires significantly more work. This is way too hard to review. You need to drastically and mercilessly simplify.

I believe it would be significantly simpler if you can introduce "allocator" interface separately without any semantic changes first. _Then_ extend that allocator with CAS allocations support.

Additionally, I think it clashes with https://github.com/openjdk/jdk/pull/31047, and that one probably needs to go in first.

kdnilsen reviewed (changes requested):
Agree with @shipilev that we should simplify this PR so that it is more accessible to more reviewers and we can assimilate the changes in smaller incremental steps.

Makes sense to start with the ShenandoahAllocator refactoring as a separate PR.

A second step might be to simplify the accounting updates that are performed immediately following each change to the freeset tallies.

A third step might be to remove alignment requirement on PLABs given upstream changes that have already been integrated.  That way we wouldn't have to special case the OldCollector Allocator.

Then maybe we can integrate the CAS allocator change.

Not sure that the order I suggest above for these incremental steps represents the smoothest evolution of the software.  If you think a different order makes more sense, that's fine with me.

pengxiaolong commented:
> Agree with @shipilev that we should simplify this PR so that it is more accessible to more reviewers and we can assimilate the changes in smaller incremental steps.
> 
> Makes sense to start with the ShenandoahAllocator refactoring as a separate PR.
> 
> A second step might be to simplify the accounting updates that are performed immediately following each change to the freeset tallies.
> 
> A third step might be to remove alignment requirement on PLABs given upstream changes that have already been integrated. That way we wouldn't have to special case the OldCollector Allocator.
> 
> Then maybe we can integrate the CAS allocator change.
> 
> Not sure that the order I suggest above for these incremental steps represents the smoothest evolution of the software. If you think a different order makes more sense, that's fine with me.

Thanks! I agree with smaller PRs instead of one giant PR with all the features supporting the CAS allocator. I will break it down to multiple phases, and create PR for each phase. 
I think we don't need to keep this PR open, next PR will be just the new allocator interface(refactoring), I have been working on it.
